sample/trace_key.go

```diff
@@ -40,75 +43,194 @@ func newTraceKey(fields []string, useTraceLength bool) *traceKey {
-	fieldCollector := make(map[string][]string)
 
 	// for each field, for each span, get the value of that field
 	spans := trace.GetSpans()
-	uniques := generics.NewSetWithCapacity[string](maxKeyLength)
+	uniques := d.distinctValue
+	uniques.init(d.fields, maxKeyLength)
+	defer func() {
+		// reset the distinctValue for reuse
+		uniques.Reset()
+		d.keyBuilder.Reset()
+	}()
 outer:
-	for _, field := range d.fields {
+	for i, field := range d.fields {
 		for _, span := range spans {
 			if val, ok := span.Data[field]; ok {
-				u := fmt.Sprintf("%s/%v", field, val)
 				// don't bother to add it if we've already seen it
-				if uniques.Contains(u) {
-					continue
-				}
-				uniques.Add(u)
-				if len(uniques) >= maxKeyLength {
+				if uniques.totalUniqueCount >= maxKeyLength {
 					break outer
 				}
-				fieldCollector[field] = append(fieldCollector[field], fmt.Sprintf("%v", val))
+				if uniques.AddAsString(val, i) {
+					continue
+				}
 			}
 		}
 	}
 	// ok, now we have a map of fields to a list of all unique values for that field.
 	// (unless it was huge, in which case we have a bunch of them)
 
-	var key strings.Builder
-	for _, field := range d.fields {
+	// please change this so that strings.Builder can be reused between calls
+
+	for i := range d.fields {
+		values := uniques.Values(i)
 		// if there's no values for this field, skip it
-		values, ok := fieldCollector[field]
-		if !ok || len(values) == 0 {
+		if len(values) == 0 {
 			continue
 		}
-		// sort and collapse list
-		sort.Strings(values)
 		var prevStr string
 		for _, str := range values {
 			if str != prevStr {
-				key.WriteString(str)
-				key.WriteRune('•')
+				d.keyBuilder.WriteString(str)
+				d.keyBuilder.WriteRune('•')
 				fieldCount += 1
 			}
 			prevStr = str
 		}
 		// get ready for the next element
-		key.WriteRune(',')
+		d.keyBuilder.WriteRune(',')
 	}
 
 	if trace.RootSpan != nil {
 		for _, field := range d.rootOnlyFields {
 			if val, ok := trace.RootSpan.Data[field]; ok {
-				key.WriteString(fmt.Sprintf("%v,", val))
+				d.keyBuilder.WriteString(fmt.Sprintf("%v,", val))
 				fieldCount += 1
 			}
 		}
 	}
 
 	if d.useTraceLength {
-		key.WriteString(strconv.FormatInt(int64(len(spans)), 10))
+		d.keyBuilder.WriteString(strconv.FormatInt(int64(len(spans)), 10))
 		fieldCount += 1
 	}
 
-	return key.String(), fieldCount
+	return d.keyBuilder.String(), fieldCount
+}
+
+// distinctValue keeps track of distinct values for a set of fields.
+// It stores the unique values as strings.
+type distinctValue struct {
+	buf          []byte
+	fields       []string
+	values       []map[uint64]string
+	valuesBuffer []string
+
+	// totalUniqueCount keeps track of how many unique values we've seen so far for a trace key.
+	totalUniqueCount int
+	// maxDistinctValue is the maximum number of distinct values we will store for a trace key.
+	maxDistinctValue int
+}
+
+func (d *distinctValue) init(fields []string, maxDistinctValue int) {
```

Comment:
This is way more complicated than it needs to be. It looks like `distinctValue.fields` is read-only, which means you don't need your own copy at all, and should just be able to do `d.fields = fields`. (I really, really wish go had immutables to make this explicit - they did it for strings but refuse to do it for other types)

Also, when you _do_ need a local copy of a slice in a pre-allocated buffer, you can just do
```
mySlice = append(mySlice[:0], theirSlice...)
```
instead of all this.